### DISCUSSION TITLE
Is it possible to hide the locale in urls for localized routing?

### DISCUSSION THREAD

**Discussion by MartinXPN** [vercel/next.js#39746] [Help] [answered]
## The Question
I was wondering if there is any way of having the same visible URL for the client while being able to handle internationalization. 

## My use case
* I want my users to have universal access to pages and hide the locales from the URL path. If a user shares a link with another user, the content should be loaded in the preferred language of the current user that opens the page.
* I'm fetching localized data on the server-side with `getServerSideProps` and caching the result with `Cache-Control` for 5-10 minutes. So, passing locale as a query parameter (`?lang=en`, `?lang=ca`, etc) probably won't work as the cached results should be different for different locales.
* It would be great to have a language-specific `Cache-Control` if there is a way of sending the language somehow in the request to `getServerSideProps`.

## What I have done before
* My app was written entirely in React and I'm now moving to Next.js. I used to pass the locale in a URL query parameter like `?lang=en`. I would then store that locale in the `localStorage`, send API requests with the given locale, and change the app's appearance accordingly.
* That used to work fine, but now I want to make use of the server-side rendering and request caching, which makes things more complicated.

-----------

This seems to be a common use case. Is there any way to handle the internationalization without explicitly changing the URL? Or maybe having one URL at first and then hiding some parts after? Any kind of idea/workaround would be really appreciated!

**Comment by gaston-flores** [comment, marked as answer]
>It would be great to have a language-specific Cache-Control if there is a way of sending the language somehow in the request to getServerSideProps.

You can get the user's preferred locales through the `Accept-Language` header on the request, and use the `Vary` header in your response to tell the cache to use `Accept-Language` as part of the cache key.

```js
import Negotiator from "negotiator"; // You don't need this library to parse the header, but it's convenient

const AVAILABLE_LOCALES = ["en", "es"];

export function getServerSideProps({ req, res }) {
  const negotiator = new Negotiator(req);
  const language = negotiator.language(AVAILABLE_LOCALES) || AVAILABLE_LOCALES[0];
  res.setHeader("Cache-Control", "max-age=120");
  res.setHeader("Vary", "Accept-Language");

  return {
    props: {
      language
    }
  };
}

export default function IndexPage({ language }) {
  return <h1>Your preferred language: {language}</h1>;
}
```

You don't need Next.js i18n routing to do this.

**Reply by MartinXPN** [reply]
Thanks a lot! That makes sense. 
And what would be the best way to set the locale from the client side to make sure it's sent to the server on each request? (My app allows the users to choose locales from within the website).

**Reply by gaston-flores** [reply]
You can do something similar to what Next does and [create a cookie to override the Accept-Language header](https://nextjs.org/docs/advanced-features/i18n-routing#leveraging-the-next_locale-cookie). Not sure how that would mix with your cache strategy though 🤔 It might be easier to use the i18n routing in that case.

**Reply by MartinXPN** [reply]
Okay I'll try to modify the `NEXT_LOCALE` cookie, and if it doesn't work out, I'll revert to i18n routing.
Thanks a lot for the tips!

**Reply by MartinXPN** [reply]
Hi, @gaston-flores I tried the suggested solution by setting the `NEXT_LOCALE` cookie on the client-side and using the code above on the `getServerSideProps`. Yet, unfortunately, seems like the `NEXT_LOCALE` only works with `i18n`-specific routing.

The `negotiator` wasn't able to capture the language which would result in the default language.

I guess I'll have to use the `i18n` routing for localization.

**Reply by gaston-flores** [reply]
Oh, you need to do a bit more legwork there. Negotiator won't know about that cookie, so you need to build that logic yourself:

```js
function getBestLanguage(availableLocales, req) {
  const { cookies } = req;
  if (cookies.LANG && availableLocales.includes(cookies.LANG)) {
    return cookies.LANG;
  }
  const negotiator = new Negotiator(req);
  return negotiator.language(availableLocales) || availableLocales[0];
}

const AVAILABLE_LOCALES = ["en", "es"];

export function getServerSideProps({ req, res }) {
  const language = getBestLanguage(AVAILABLE_LOCALES, req);

  return {
    props: {
      language
    }
  };
}
```

Your `LANG` cookie should override the accept header in this case.
